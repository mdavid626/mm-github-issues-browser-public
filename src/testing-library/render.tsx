/* istanbul ignore file */
import { InitialEntry, Location } from '@remix-run/router';
import { Queries, queries } from '@testing-library/dom';
import {
  RenderHookOptions,
  RenderOptions,
  render,
  renderHook,
  RenderResult,
  RenderHookResult,
} from '@testing-library/react';
import React, { ReactElement } from 'react';
import { MemoryRouter, useLocation, useNavigate } from 'react-router-dom';
import { NavigateFunction } from 'react-router/dist/lib/hooks';

type SimpleRouter = {
  location?: Location;
  navigate?: NavigateFunction;
};

type RouterResult = {
  router: SimpleRouter;
};

export const renderWithRouter = (
  component: ReactElement,
  options?: Omit<RenderOptions, 'queries'>,
  initialEntries?: InitialEntry[]
): RenderResult & RouterResult => {
  const router = {} as SimpleRouter;
  return {
    ...render(component, {
      wrapper: ({ children }) => {
        const Children = () => {
          router.location = useLocation();
          router.navigate = useNavigate();
          return children;
        };
        return (
          <MemoryRouter initialEntries={initialEntries}>
            <Children />
          </MemoryRouter>
        );
      },
      ...options,
    }),
    router,
  } as RenderResult & RouterResult;
};

export const renderHookWithRouter = <
  Result,
  Props,
  Q extends Queries = typeof queries,
  Container extends Element | DocumentFragment = HTMLElement,
  BaseElement extends Element | DocumentFragment = Container
>(
  render: (initialProps: Props) => Result,
  options?: RenderHookOptions<Props, Q, Container, BaseElement>,
  initialEntries?: InitialEntry[]
): RenderHookResult<Result, Props> & RouterResult => {
  const router = {} as SimpleRouter;
  return {
    ...renderHook(render, {
      wrapper: ({ children }) => {
        const Children = () => {
          router.location = useLocation();
          router.navigate = useNavigate();
          return children;
        };
        return (
          <MemoryRouter initialEntries={initialEntries}>
            <Children />
          </MemoryRouter>
        );
      },
      ...options,
    }),
    router,
  };
};
