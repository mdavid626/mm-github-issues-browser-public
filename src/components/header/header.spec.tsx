import { cleanup, screen } from '@testing-library/react';
import React from 'react';
import { renderWithRouter } from '../../testing-library/render';
import Header from './header';

describe('header', () => {
  afterEach(cleanup);

  it('should render', () => {
    const { asFragment } = renderWithRouter(<Header />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('should highlight about page when on about page', () => {
    renderWithRouter(<Header />, undefined, ['/about']);
    expect(screen.getByText('React GitHub Issues')).not.toHaveClass(
      'Header-selectedLink'
    );
    expect(screen.getByText('About')).toHaveClass('Header-selectedLink');
  });

  it('should highlight issues page when on issue detail page', () => {
    renderWithRouter(<Header />, undefined, ['/issue/1']);
    expect(screen.getByText('React GitHub Issues')).toHaveClass(
      'Header-selectedLink'
    );
    expect(screen.getByText('About')).not.toHaveClass('Header-selectedLink');
  });
});
