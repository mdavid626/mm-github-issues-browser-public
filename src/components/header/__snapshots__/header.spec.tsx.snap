// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`header should render 1`] = `
<DocumentFragment>
  <div
    class="Header"
  >
    <div
      class="Header-title"
    >
      GitHub Issues Browser
    </div>
    <div
      class="Header-navigation"
    >
      <a
        class="Header-link Header-selectedLink"
        href="/"
      >
        React GitHub Issues
      </a>
      <a
        class="Header-link"
        href="/about"
      >
        About
      </a>
    </div>
  </div>
</DocumentFragment>
`;
